### PR TITLE
Rename color constants

### PR DESCRIPTION
**Pull request by Vivraan**
Fix naming of certain constants and color name strings in C++ and C#.

Advise on how to parse constants in UPPERCASE_NAMING style for gdscript Color Constants.

### CODE REVIEW

**Comment by bojidar-bg** [comment]
> Advise on how to parse constants in UPPERCASE_NAMING style for gdscript Color Constants.

Color constants are registered here:

https://github.com/godotengine/godot/blob/ec678c5e94ac1c91c099c15fabc398e80ede7bfe/core/variant_call.cpp#L2334-L2337

In order for the color name to be uppercase, the second argument to `add_variant_constant` (currently `color->key()`) should be an uppercase string.

To do this you might (a) modify `_populate_colors()` to use upper-case color names or (b) use [`String::to_upper`](https://github.com/godotengine/godot/blob/ec678c5e94ac1c91c099c15fabc398e80ede7bfe/core/ustring.cpp#L968) to transform the color name.

**Comment by ShivamMukherjee** [comment]
I just added the entire bunch of names as new entries. Old code needn't break.

**Comment by ShivamMukherjee** [comment]
However this doesn't follow the consistency that @aaronfranke wanted to provide (`Colors.Red` in C# and `Color.RED` in GDScript).

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs [hunk 1]
```diff
@@ -171,12 +172,14 @@ public static class Colors
         public static Color Blue { get { return namedColors["blue"]; } }
         public static Color BlueViolet { get { return namedColors["blueviolet"]; } }
         public static Color Brown { get { return namedColors["brown"]; } }
-        public static Color BurlyWood { get { return namedColors["burlywood"]; } }
+        public static Color BurlyWood { get { return namedColors["burlywood"]; } } // for compatibility
+        public static Color Burlywood { get { return namedColors["burlywood"]; } }
```
Comment (aaronfranke) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs:176]: It should be kept as `BurlyWood`. The two words stand fine on their own, and "Burlywood" isn't a very common word, but most importantly, it matches the X11 and .NET color names.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs [hunk 2]
```diff
@@ -280,7 +284,8 @@ public static class Colors
         public static Color Salmon { get { return namedColors["salmon"]; } }
         public static Color SandyBrown { get { return namedColors["sandybrown"]; } }
         public static Color SeaGreen { get { return namedColors["seagreen"]; } }
-        public static Color SeaShell { get { return namedColors["seashell"]; } }
+        public static Color SeaShell { get { return namedColors["seashell"]; } } // for compatibility
+        public static Color Seashell { get { return namedColors["seashell"]; } }
```
Comment (aaronfranke) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs:288]: Same as above.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs [hunk 3]
```diff
@@ -171,12 +172,14 @@ public static class Colors
...
-        public static Color BurlyWood { get { return namedColors["burlywood"]; } }
+        public static Color BurlyWood { get { return namedColors["burlywood"]; } } // for compatibility
+        public static Color Burlywood { get { return namedColors["burlywood"]; } }
         public static Color CadetBlue { get { return namedColors["cadetblue"]; } }
         public static Color Chartreuse { get { return namedColors["chartreuse"]; } }
         public static Color Chocolate { get { return namedColors["chocolate"]; } }
         public static Color Coral { get { return namedColors["coral"]; } }
-        public static Color Cornflower { get { return namedColors["cornflower"]; } }
+        public static Color Cornflower { get { return namedColors["cornflower"]; } } // for compatibility
+        public static Color CornflowerBlue { get { return namedColors["cornflowerblue"]; } }
```
Comment (aaronfranke) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs:182]: Good catch with `CornflowerBlue` and `LimeGreen`, those are the names in both the X11 and .NET color names. However, we don't need to keep the old names, we can break compatibility for 4.0.

**Review by aaronfranke — CHANGES_REQUESTED** [review]

**Comment by aaronfranke** [comment]
> I don't see why `Colors.RED` or for that matter all of these constants can't be added in as a plugin instead

They could be, but this is a very simple feature so it's easy enough to just have it built-in, and arguably commonly used (the colors are used less othen than the constants like Vector2.UP, but they are both sets of standardized constants so we should treat them the same).

When it comes to `Color.RED` vs `Colors.RED`, the former isn't terrible, but IMO with the sheer amount of constants it makes sense to separate them. In C# there is also the concern of `Color.Red` etc obscuring unrelated methods like `Color.Color8`, and in GDScript these are just defined as global methods, but perhaps they should be moved to `Color` in GDScript too (and if so, that's a strong case for `Colors.RED`).

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs (hunk 1 above)
Comment (ShivamMukherjee) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs:176]: Wikipedia is incorrect? Welp, not surprised.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs (hunk 3 above)
Comment (ShivamMukherjee) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs:182]: I observed all the PRs related to the breaking compatibility bits, so I'll comply.

Comment (ShivamMukherjee) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs:182]: What about the ones in [color_names.inc]?(https://github.com/godotengine/godot/pull/41019/files/3b118080217bddafee350aeb6cb37ac687e6f7ce#diff-6caad16a9471f9f3b427c06ce47a9b78)

**Comment by ShivamMukherjee** [comment]
> When it comes to `Color.RED` vs `Colors.RED`, the former isn't terrible, but IMO with the sheer amount of constants it makes sense to separate them. In C# there is also the concern of `Color.Red` etc obscuring unrelated methods like `Color.Color8`, and in GDScript these are just defined as global methods, but perhaps they should be moved to `Color` in GDScript too (and if so, that's a strong case for `Colors.RED`).

Would this be done by adding another Variant Type in [variant_call.cpp](https://github.com/godotengine/godot/blob/ec678c5e94ac1c91c099c15fabc398e80ede7bfe/core/variant_call.cpp#L2334-L2337)?

File: doc/classes/Color.xml [hunk 4]
```diff
@@ -348,7 +348,7 @@
 			Brown color.
 		</constant>
 		<constant name="burlywood" value="Color( 0.87, 0.72, 0.53, 1 )">
-			Burly wood color.
+			Burlywood color.
```
Comment (aaronfranke) [inline doc/classes/Color.xml:351]: Should be "Burly wood"

File: doc/classes/Color.xml [hunk 5]
```diff
@@ -362,8 +362,8 @@
 		<constant name="coral" value="Color( 1, 0.5, 0.31, 1 )">
 			Coral color.
 		</constant>
-		<constant name="cornflower" value="Color( 0.39, 0.58, 0.93, 1 )">
-			Cornflower color.
+		<constant name="cornflowerblue" value="Color( 0.39, 0.58, 0.93, 1 )">
+			Cornflower Blue color.
```
Comment (aaronfranke) [inline doc/classes/Color.xml:366]: Should be "Cornflower blue".

**Review by aaronfranke — CHANGES_REQUESTED** [review]
As for the constants, I'm not sure of the best way to implement them. Take a look at the PR that added them. If you can't figure something out, I'll try myself later.

File: core/color_names.inc [hunk 6]
```diff
@@ -152,4 +152,152 @@ static void _populate_named_colors() {
 	_named_colors.insert("whitesmoke", Color(0.96, 0.96, 0.96));
 	_named_colors.insert("yellow", Color(1.00, 1.00, 0.00));
 	_named_colors.insert("yellowgreen", Color(0.60, 0.80, 0.20));
+
+	// You're a pedant
```
[the remark this answers is not in the document]
Comment (ShivamMukherjee) [reply core/color_names.inc:156]: Only a pedantic person would go to any length to introduce constants that follow the rule.

Dropping it.

File: core/color_names.inc [hunk 7]
```diff
@@ -23,7 +23,7 @@ static void _populate_named_colors() {
 	_named_colors.insert("chartreuse", Color(0.50, 1.00, 0.00));
 	_named_colors.insert("chocolate", Color(0.82, 0.41, 0.12));
 	_named_colors.insert("coral", Color(1.00, 0.50, 0.31));
-	_named_colors.insert("cornflower", Color(0.39, 0.58, 0.93));
+	_named_colors.insert("cornflowerblue", Color(0.39, 0.58, 0.93));
```
[the remark this answers is not in the document]
Comment (ShivamMukherjee) [reply core/color_names.inc:26]: In [color.cpp](https://github.com/godotengine/godot/blob/master/core/color.cpp), `Color::named` normalises the input to favour all lowercase and no spaces. This should work even if the new uppercase constants are in a new map used solely for regstering constants in `variant_call`.

File: doc/classes/Color.xml (hunk 4 above)
Comment (ShivamMukherjee) [reply doc/classes/Color.xml:351]: Favouring Burlywood here since it's the real word.

File: doc/classes/Color.xml (hunk 5 above)
Comment (ShivamMukherjee) [reply doc/classes/Color.xml:366]: Right, keeping with the rest of the comments.

File: doc/classes/Color.xml (hunk 4 above)
Comment (ShivamMukherjee) [reply doc/classes/Color.xml:351]: I'll go with Burly wood for consistency.

File: core/color_names.inc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShivamMukherjee) [reply core/color_names.inc:26]: That said, it's probably okay to remove the lowercase names from the docs.

[the remark this answers is not in the document]
Comment (aaronfranke) [reply core/color_names.inc:26]: The constants should be uppercase, and we should find a way to do this without duplicating every single color in `_named_colors`.

[the remark this answers is not in the document]
Comment (ShivamMukherjee) [reply core/color_names.inc:28]: I created a new map for the uppercase names. We can now decide on whether to change the normalisation rule or not.

**Comment by bojidar-bg** [comment]
Instead of duplicating the color name list, wouldn't it be easier to change this:
```c++
_VariantCall::add_variant_constant(Variant::COLOR, color->key(), color->value());
```
to this:
```c++
_VariantCall::add_variant_constant(Variant::COLOR, color->key().to_upper(), color->value());
```

**Comment by ShivamMukherjee** [comment]
> Instead of duplicating the color name list, wouldn't it be easier to change this:
> 
> ```c++
> _VariantCall::add_variant_constant(Variant::COLOR, color->key(), color->value());
> ```
> 
> to this:
> 
> ```c++
> _VariantCall::add_variant_constant(Variant::COLOR, color->key().to_upper(), color->value());
> ```

No, it would miss out on the added underscores that denote spacing.

**Comment by ShivamMukherjee** [comment]
> As for the constants, I'm not sure of the best way to implement them. Take a look at the PR that added them. If you can't figure something out, I'll try myself later.

I believe we'll need to remove `ColorN` and introduce a `Colors` built-in dictionary/variant type which supports the `in` operator, if needed.

File: core/color_names.inc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShivamMukherjee) [reply core/color_names.inc:28]: I duplicated the map and used the new map in `_VariantCall`, which is causing linker errors in CI.

**Comment by ShivamMukherjee** [comment]
> As for the constants, I'm not sure of the best way to implement them. Take a look at the PR that added them. If you can't figure something out, I'll try myself later.

Do you mean adding a new paceholder for the Color constants or just updating them in the editor? The latter is done by 
https://github.com/godotengine/godot/blob/ec678c5e94ac1c91c099c15fabc398e80ede7bfe/core/variant_call.cpp#L2334-L2337

**Comment by ShivamMukherjee** [comment]
Any updates on this?

File: core/color.cpp [hunk 8]
```diff
@@ -398,12 +398,17 @@ Color Color::named(const String &p_name) {
...
 	// Normalize name
-	name = name.replace(" ", "");
-	name = name.replace("-", "");
-	name = name.replace("_", "");
+	// name = name.replace(" ", "");
+	// name = name.replace("-", "");
+	// name = name.replace("_", "");
+	// name = name.replace("'", "");
+	// name = name.replace(".", "");
+	// name = name.to_lower();
```
Comment (bojidar-bg) [inline core/color.cpp:406]: Remove commented-out code, people wanting to see the history can always check it through `git`.

Comment (ShivamMukherjee) [reply core/color.cpp:406]: Of course.

**Comment by bojidar-bg** [comment]
Overall, looks good to me.

**Comment by ShivamMukherjee** [comment]
I think if we do attempt to do `Colors.RED` we'll need a new `Variant` class to hold the constants, or we could end up making the constants global (not the intended effect of this PR, and although following the design of the Error codes and related constants, in sharp contrast with the C# implementation).

**Comment by ShivamMukherjee** [comment]
Should we move forward this then?

**Comment by aaronfranke** [comment]
You forgot about C#'s `ColorN`. The behavior should match GDScript's `ColorN`. I'm also concerned about `ColorN` requiring underscores, since that would mean `"seagreen"` doesn't work anymore which is fine I guess, but also `"SeaGreen"` wouldn't work either which isn't good because that's the property name in C# so it should work.

Perhaps we would need some kind of mapping system so that `ColorN` strips out underscores and looks for either `SEAGREEN` or `seagreen` in the dictionary, but the exposed constants can have slightly different names like `SEA_GREEN`.

**Comment by ShivamMukherjee** [comment]
`ColorN` uses `Color::named` underneath. I'll need to change the normalisation rule back to the previous one for this to work.

**Comment by ShivamMukherjee** [comment]
Why don't we use this to disambiguate spelling errors in Color names?

https://github.com/godotengine/godot/blob/a6916d1f179f760f015f385c01729d205af97abe/core/ustring.cpp#L2734

**Review by Vivraan — COMMENTED** [review]
This implements the third case I described. It'll probably fail on the link tests, and it has an out_of_bounds error.

No new branches and tests included.

Doesn't work btw, but I suppose the intent is clear.

**Review by Vivraan — COMMENTED** [review]
This is the correct implementation. Please check if all the generated names are correct or not (I've done a preliminary check on my own machine).

There are linker errors wherever `variant_call.cpp` is not linked through the build command.

**Comment by ShivamMukherjee** [comment]
@bojidar-bg @aaronfranke any way to let the linker know that my new function is being used and needed to be linked?

**Comment by ShivamMukherjee** [comment]
I kinda believe that this implementation can make its way to 3.2.

**Comment by Calinou** [comment]
@Vivraan This pull request breaks backwards compatibility since it *renames* color constants. Old scripts that rely on those constants will cease to work until they've been fixed. Therefore, we won't cherry-pick this pull request to the `3.2` branch if it's merged in `master`.

**Comment by Vivraan** [comment]
Yeah I missed that when I made the comment. I'm fine with it going to 4.0.

However, one of these commits consists of using both uppercase and lowercase constants, and can be adopted into 3.2 if needed.

**Comment by Vivraan** [comment]
Every build test apart from Static and Windows is failing because they're treating `_to_uppercase_name` as unused where it is called in `variant_call.cpp`. Could someone please look into this?

**Comment by akien-mga** [comment]
> Every build test apart from Static and Windows is failing because they're treating `_to_uppercase_name` as unused where it is called in `variant_call.cpp`. Could someone please look into this?

I guess it fails because you made it static, so it's copied into `color.cpp`, but when compiling `color.cpp` it's unused.

**Comment by Vivraan** [comment]
That solves the linker errors.

Now, the last thorn:
```
DEEP_SK_Y_BLUE = Color( 0, 0.75, 1, 1 )
```
is the only name which isn't coming up properly - should I add this to the edge cases list?

**Review by Vivraan — COMMENTED** [review]
This is the version for 4.0 - breaks backwards compatibility,

**Review by Vivraan — COMMENTED** [review]
~~This is the version for 3.2 - fully backwards compatible.~~
Bad comment, should be on [65d3e2e](https://github.com/godotengine/godot/pull/41019/commits/65d3e2ef530f2d59c554720fe1dadd9373761708).

This is, of course, going to break code which uses `cornflower` instead of the correct `cornflowerblue`.

**Comment by Vivraan** [comment]
Can we move forward this as it is right now?

**Comment by aaronfranke** [comment]
@Vivraan Be patient, some PRs sit waiting for years. I haven't been at my main computer for the past 2 weeks, and my current one can't run Vulkan, so I haven't been able to test this yet or do a proper review.

**Comment by Vivraan** [comment]
Indeed, this is a cosmetic change, it can wait.

**Comment by Vivraan** [comment]
Requesting review.

**Comment by aaronfranke** [comment]
@Vivraan Sorry for the delay, I forgot about this. Anyway, this needs to be rebased on the latest master before it can be reviewed.

File: core/uppercase_name_transformation.inc [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+		_modifiable_colors.append("TURQUOISE");
+		_modifiable_colors.append("WHITE");
+		_modifiable_colors.append("YELLOW");
+	}
+
+	if (_color_descriptors.empty()) {
+		_color_descriptors.append("DARK");
+		_color_descriptors.append("MEDIUM");
+		_color_descriptors.append("LIGHT");
+		_color_descriptors.append("OLIVE");
```
[the remark this answers is not in the document]
Comment (Vivraan) [reply core/uppercase_name_transformation.inc:35]: Yeah, it's sensible to describe the hardcoded list as "this prevents looping extra for less frequently occurring colour names." There simply isn't enough use of Lime apart from Lime Green(?) and for Navy with Navy Blue, however Olive appears for some more colours as well.

Good catch. I will rectify this quickly.

**Comment by Vivraan** [comment]
Should I reintroduce the old names for 3.2 compat? I guess a commit with that can go into the 3.2 branch but one without it can go into 4.0 only.

File: core/uppercase_name_transformation.inc (hunk 9 above)
[the remark this answers is not in the document]
Comment (Vivraan) [reply core/uppercase_name_transformation.inc:35]: Oh, just figured that Olive is very infrequent as well, and has already been put in the edge case if-ladder. I'll remove this item.

**Comment by Vivraan** [comment]
> @Vivraan Sorry for the delay, I forgot about this. Anyway, this needs to be rebased on the latest master before it can be reviewed.

I'm confused about the files I'm modifying not being there in the master branch. Could you confirm if they have been removed or not?

**Comment by Vivraan** [comment]
@reduz could you review this? Seems you were the author of the moving change.

**Comment by Vivraan** [comment]
> As for the constants, I'm not sure of the best way to implement them. Take a look at the PR that added them. If you can't figure something out, I'll try myself later.

I believe @reduz has addressed this in the latest changes relating to code reorganisation.

**Comment by Vivraan** [comment]
Looks like my transformation code is wasteful. I'll need to provide the names directly in color_names.inc, or wherever the colour names are registered.

**Comment by aaronfranke** [comment]
@Vivraan This can't be done to 3.2 because it breaks compatibility.

What you have here is a merge, but you need a rebase (and ideally squash). See [this article](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html) for more information.

**Comment by Vivraan** [comment]
> @Vivraan This can't be done to 3.2 because it breaks compatibility.
> 
> What you have here is a merge, but you need a rebase and ideally squash, where it says "Commits 29" it should say "Commits 1". See [this article](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html) for more information.

It looks like this because the files I was trying to change have been (re)moved. Atm I'm waiting for Reduz to review this change, since it looks like the way these constants are defined or introduced has been completely redone.

**Comment by dalexeev** [comment]
@Vivraan I opened alternative PR #45144 and listed you as a co-author.

**Comment by Vivraan** [comment]
> @Vivraan I opened alternative PR #45144 and listed you as a co-author.

Amazing! I'm glad the named corrections have been updated to the latest API!
Are the C# colours handled in the same way as GDScript or are they named in `PascalCase`?

**Comment by dalexeev** [comment]
> Are the C# colours handled in the same way as GDScript or are they named in `PascalCase`?

GDScript: `Color.LIGHT_SEA_GREEN`
C#: `Colors.LightSeaGreen`

**Comment by aaronfranke** [comment]
For the record: #45144, which replaces this PR, has been merged.

**Pull request closed without merging**